accepted/install-locations.md

```diff
@@ -0,0 +1,241 @@
+
+The problem already exists since the `apphost` has been shipped in .NET Core 2.2 with support for framework dependent apps. The feature is used only occasionally. The most common case is probably dotnet global tools (`dotnet tool install -g <tool>`) which create framework dependent `apphost` for each installed tool.
+
+In .NET Core 3.0 framework dependent apps use an `apphost` by default, which makes this case far more prevalent. Scenarios requiring the other hosts are also likely to be relatively common (COM, IJW, native hosting). So the hosts which we ship with .NET Core 3.0 will be spread across all machines running .NET Core (not just development but production as well) without any ability to service them directly.
+
+This means that .NET Core 3.0 is the last chance to get the algorithms for install locations modified in "breaking ways". Future release will very likely have to be 100% backward compatible in this regard.
+
+
+## Proposal for 3.0
+
+### Globally registered install location :new:
+This is by far the most common case and in 3.0 we want to provide a way to globally install to a custom location. The typical case will still be to install into the default location, but custom location should be possible. Either location should behave the same and which one is used is solely decision of the installer.
+
+Note that the installer should try to avoid creating two global locations as only one can be registered. So if there is already some version installed it should only add to that location and create a new one. The only way to get two global locations should be to first install 3.0 into custom location and then install 2.* which always installs to the default location.
+
+To achieve the desired behavior the install location must be recorded in some system-wide configuration in a fixed location which is well-known.
+
+* Windows - the installer will store the install location in 32-bit registry (for both 32-bit and 64-bit installations):
+  * `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` - where `<arch>` is one of the `x86`, `x64`, `arm` or `arm64`.
+  Note that this registry key is "redirected" that means that 32-bit processes see different copy of the key then 64bit processes. So it's important that both installers and the host access only the 32-bit view of the registry.
+* Linux and macOS - the installer will store the install location in a text file
+  * `/etc/dotnet/install.conf` - The file will be in the `.ini` [format](https://en.wikipedia.org/wiki/INI_file). In it the key `InstallLocation` (in no section) will contain the path to the global location. The host will have a simplistic parser for this file, which will only support empty lines, comment lines (starts with #) and the one key/value pair. Any additional keys will be ignored. The host will trim any spaces around the equal sign and then treat the rest of the line after the equal sign as the install location path (no quotes or escaping).
```

Comment:
Maybe use a different name than `install.conf` so the file can be used for other things (and avoid the host having to read a bunch of files for different settings). suggestion: `/etc/dotnet/config`

Reply:
I'm leaning towards @nguerrera argumentation that the file which will be read by the hosts (`apphost` most commonly) should be as simple as possible. So a simple text file with one line which is the path, nothing else in it. The host would only read the first line and ignore the rest, so that there's at least some extensibility, but no fancy syntax. Adding even a basic `ini` parser into all the hosts is risky:
* It's a parser - so security sensitive - although the severity of this is not too high
* Non-trivial amount of code - size is a consideration
* No way to service it - these are part of the app not the .NET Core install
* Unless it implements a full `ini` "spec" (whatever that means), there's a risk that in the future we would run into its limitations without an ability to fix it.

If its purpose is really only to contain the install location, then I think the previously discussed: `/etc/dotnet/install_location` (as a file with no extension) would work pretty well.

I would like to avoid overly generic things like `config` and so on, as it would be easy to expect it to contain lot of other configuration options.

As for the other settings (if and when we introduce a configuration file for them) - they would be read by a different component in the system (probably either `hostfxr` or `hostpolicy` or even `dotnet.dll`). All of those are much easier to service and are shared, so including a more complex piece of code in them is not as risky. For the same reason the necessity to read a separate file is not really a downside as the different components would have to read it separately anyway (even if it would be the same file).

Reply:
Fwiw we had a startup file for a bit (not shipped) that specified the "approot". At that time we thought a JSON would be the most flexible if we ever needed it, although I believe we brought in the casablanca JSON parser which added a bit more size to the exe. Here's the commit that removed it: https://github.com/dotnet/core-setup/commit/d6802d0c09f845cc26020691f155320a1fb37a67#diff-d879441c77c53723780bea5613c6056d